core/src/main/java/org/apache/iceberg/io/BaseDeltaWriter.java

```diff
@@ -0,0 +1,248 @@
+
+  private final PositionDelete<T> positionDelete = new PositionDelete<>();
+  private final StructLikeMap<List<FilePos>> insertedRowMap;
+
+  // Function to convert the generic data to a StructLike.
+  private final Function<T, StructLike> structLikeFun;
+
+  public BaseDeltaWriter(RollingContentFileWriter<DataFile, T> dataWriter) {
+    this(dataWriter, null);
+  }
+
+  public BaseDeltaWriter(RollingContentFileWriter<DataFile, T> dataWriter,
+                         RollingContentFileWriter<DeleteFile, PositionDelete<T>> posDeleteWriter) {
+    this(dataWriter, posDeleteWriter, null, null, null, null);
+  }
+
+  public BaseDeltaWriter(RollingContentFileWriter<DataFile, T> dataWriter,
+                         RollingContentFileWriter<DeleteFile, PositionDelete<T>> posDeleteWriter,
+                         RollingContentFileWriter<DeleteFile, T> equalityDeleteWriter,
+                         Schema tableSchema,
+                         List<Integer> equalityFieldIds,
+                         Function<T, StructLike> structLikeFun) {
+
+    Preconditions.checkNotNull(dataWriter, "Data writer should always not be null.");
+
+    if (posDeleteWriter == null) {
+      // Only accept INSERT records.
+      Preconditions.checkArgument(equalityDeleteWriter == null);
+    }
+
+    if (posDeleteWriter != null && equalityDeleteWriter == null) {
+      // Only accept INSERT records and position deletion.
+      Preconditions.checkArgument(tableSchema == null);
+      Preconditions.checkArgument(equalityFieldIds == null);
+    }
+
+    if (equalityDeleteWriter != null) {
+      // Accept insert records, position deletion, equality deletions.
+      Preconditions.checkNotNull(posDeleteWriter,
+          "Position delete writer shouldn't be null when writing equality deletions.");
+      Preconditions.checkNotNull(tableSchema, "Iceberg table schema shouldn't be null");
+      Preconditions.checkNotNull(equalityFieldIds, "Equality field ids shouldn't be null");
+      Preconditions.checkNotNull(structLikeFun, "StructLike function shouldn't be null");
+
+      Schema deleteSchema = TypeUtil.select(tableSchema, Sets.newHashSet(equalityFieldIds));
+      this.insertedRowMap = StructLikeMap.create(deleteSchema.asStruct());
+      this.structLikeFun = structLikeFun;
+    } else {
+      this.insertedRowMap = null;
+      this.structLikeFun = null;
+    }
+
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.posDeleteWriter = posDeleteWriter;
+  }
+
+  @Override
+  public void writeRow(T row) throws IOException {
+    if (enableEqualityDelete()) {
+      FilePos filePos = FilePos.create(dataWriter.currentPath(), dataWriter.currentPos());
+      insertedRowMap.compute(structLikeFun.apply(row), (k, v) -> {
+        if (v == null) {
+          return Lists.newArrayList(filePos);
+        } else {
+          v.add(filePos);
+          return v;
```

Comment:
Do we expect multiple inserts for the same key? I would expect that we do not.

If that's the case, then the suggestion below to use `StructLikeMap<FilePos>` could simplify this logic. When a row is deleted by position (added to `deletePositions`), it could be removed from the map. Then this logic could be:

```java
  StructLike key = structLikeFun.apply(row);
  FilePos previous = insertedRowMap.put(key, filePos);
  ValidationException.check(previous == null, "Detected duplicate insert for %s", key);
```

Reply:
Let's consider the three cases: 

Case.1 :   Import CDC data whose table has an unique key to iceberg table.   In this case,  it's impossible to have duplicated records with the same key to INSERT.

Case.2 :  Import CDC data whose table does not have an unique key to iceberg table,  the equality fields should be all columns in the table,  but it be possible that we will encounter duplicated rows because database such as mysql would allow to write the same record twice.  But that's not a common case, more people will create a table with unique keys, I think it's OK to not address this case. 

Case.3:   upsert records in the iceberg table.  it must have primary keys. The `UPSERT` will be executed as an `DELETE` and an `INSERT`,  then we could also ensure that duplicated keys won't happen in the `writeRow` method ( which will be called when `INSERT` row).

So in general,  I think it's OK to expect that we don't have multiple inserts for the same key.

Reply:
Oh, there's another issue. Consider the two cases: 

```
Case.1 : Execute two transactions: 

txn1:  INSERT(1) ; DELETE(1); INSERT(1);
txn2: INSERT(1) ; 
```

It won't throw any duplicate key exception because we don't check the key existences cross transactions. 

```
Case.2:  Execute only one transaction 

txn1:    INSERT(1) ; INSERT(1); 
```

we will encounter a `ValidationException` when executing the second `INSERT(1)`. 

Will it confuse users because of the different behavior ? 

For me,   iceberg won't check the key existence when inserting so in theory we don't have to guarantee that iceberg must throw a duplicated key exception if users really insert two duplicated records.

Reply:
Also, now that I'm thinking about it, we should _never_ throw an exception for this data because that would break processing. Instead, we should probably send the record to some callback (default no-op), log a warning, and either delete the previous copy or ignore the duplication.

The next question is: if we do expect duplicate inserts, then what is the right behavior? Should we make the second insert replace the first? Or just ignore the duplication?

I'm leaning toward adding a delete to replace the record, but that would only work if the two inserts were in the same checkpoint. If they arrive a few minutes apart, then the data would be duplicated in the table. But, since we consider the records identical by the insert key, I think it is correct to add a position delete for the first record.

If we choose to ignore the duplication, we then need to keep track of both insert positions in case we get INSERT(1), INSERT(1), DELETE(1). The delete would need to drop both rows, not just the second. That's why I would say that a duplicate insert replaces the previously inserted row. That way we can keep track of just one FilePos per key and lower the complexity of tracking this.

Reply:
Think about it again, My conclusion is :  we should choose the solutions: 

1. When a INSERT come in, then: 
   a.     Eliminating duplicated INSERT (adding a delete to replace the record as you said) in the same checkpoint for a given TaskWriter; 
   b.     Write it into data file, and just ignore the duplication INSERTs between checkpoints; 

2. when a DELETE come in,  then: 
     a.   write the pos-delete file if key exists in the `insertedRowMap` ; 
     b.   always write the equality-delete. 


Firstly, eliminating duplicated INSERT is easy, because we've already had the `insertedRowMap`, if there's a coming duplicated insert key, then we just replace the old INSERT (by writing the pos-delete file) with new INSERT (by writing the insert data file).  The `insertedRowMap` don't have to track `List<FilePos>` per key,  only need the `FilePos` per key. 

Second,  eliminating the duplicated INSERT between checkpoints is high-cost and we don't have to. because the equality-deletions will mask all duplicated INSERT for the same key in the previous checkpoint. 


The code should be like that: 

```java
@Override
  public void writeRow(T row) {
    if (allowEqDelete()) {
      RowOffset rowOffset = RowOffset.create(dataWriter.currentPath(), dataWriter.currentRows());

      // Copy the key to avoid messing up the insertedRowMap.
      StructLike copiedKey = asCopiedKey(row);

      // Adding a pos-delete to replace the old row.
      RowOffset previous = insertedRowMap.put(copiedKey, rowOffset);
      if (previous != null) {
        posDeleteWriter.write(positionDelete.set(previous.path, previous.rowId, row));
      }
    }

    dataWriter.write(row);
  }

  @Override
  public void writeEqualityDelete(T equalityDelete) {
    Preconditions.checkState(allowEqDelete(), "Could not accept equality deletion.");

    StructLike key = asKey(equalityDelete);
    RowOffset previous = insertedRowMap.get(key);

    if (previous != null) {
      posDeleteWriter.write(positionDelete.set(previous.path, previous.rowId, equalityDelete));
      insertedRowMap.remove(key);
    }

    eqDeleteWriter.write(equalityDelete);
  }
```